File: 프로그래머스/2/49993. 스킬트리/스킬트리.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>

using namespace std;

void clear(string skill, unordered_map<char, bool> &table ){
    
    table[skill[0]] = true;
    for(int i = 1; i < skill.size(); i++){
        table[skill[i]] = false;
    }
    
}

int solution(string skill, vector<string> skill_trees) {
    int answer = 0;
    unordered_map<char, bool> table;
    
    table.insert({skill[0], true});
    for(int i = 1; i < skill.size(); i++){
        char now = skill[i];
        table.insert({now, false});
    }
    
    
    for(auto now : skill_trees){
        
        bool flag = true;
        clear(skill, table);
        for(int i = 0;i < now.size(); i++){
            char now_skill = now[i];
            
            if(table.find(now_skill) != table.end()){
                
                if(table[now_skill]){
                    for(int i = 0; i < skill.size(); i++){
                        if(skill[i] == now_skill){
                            if(i + 1 < skill.size()){
                                table[skill[i+1]] = true;
                            }
                        }
                    }
                }
                else{
                    flag = false;
                    break;
                }
            }
        }
        
        if(flag)
            answer++;        
    }
    
    
    return answer;
}

File: 프로그래머스/2/49993. 스킬트리/스킬트리.java
import java.util.*;

class Solution {
    
    public static void clear(String skill, Map<Character, Boolean> table) {
        table.put(skill.charAt(0), true);
        for (int i = 1; i < skill.length(); i++) {
            table.put(skill.charAt(i), false);
        }
    }
    
    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        Map<Character, Boolean> table = new HashMap<>();

        table.put(skill.charAt(0), true);
        for (int i = 1; i < skill.length(); i++) {
            table.put(skill.charAt(i), false);
        }

        for (String now : skill_trees) {
            boolean flag = true;
            clear(skill, table);

            for (int i = 0; i < now.length(); i++) {
                char nowSkill = now.charAt(i);

                if (table.containsKey(nowSkill)) {
                    if (table.get(nowSkill)) {
                        for (int j = 0; j < skill.length(); j++) {
                            if (skill.charAt(j) == nowSkill) {
                                if (j + 1 < skill.length()) {
                                    table.put(skill.charAt(j + 1), true);
                                }
                            }
                        }
                    } else {
                        flag = false;
                        break;
                    }
                }
            }

            if (flag) {
                answer++;
            }
        }

        return answer;
    }
}

